module/zfs/zfs_ioctl.c

```diff
@@ -4791,53 +4789,22 @@ zfs_ioc_send(zfs_cmd_t *zc)
 		dsl_pool_rele(dp, FTAG);
 	}
 
-	if (estimate) {
-		dsl_pool_t *dp;
-		dsl_dataset_t *tosnap;
-		dsl_dataset_t *fromsnap = NULL;
-
-		error = dsl_pool_hold(zc->zc_name, FTAG, &dp);
-		if (error != 0)
-			return (error);
+	if (estimate)
+		return (SET_ERROR(EINVAL)); /* no longer supported */
```

Comment:
The failure i was seeing in `zfs_send_006_pos` was caused by a kmod/libzfs version mismatch: old libzfs asks for send space estimation still using `ZFS_IOC_SEND` and the kmod dumps raw data to the terminal because `estimate` is no longer supported/considered:

```
root@linux:~# zfs send -v -n -i $POOLNAME/fs@snap1 $POOLNAME/fs@snap2
�˺�!��Y�|�1�y~�����l�&testpool/fs@snap2��P^L���j��O�0�pH�x�|�1�y~�g�I	S܋�����(��
                                                                                           Jh�'send from @snap1 to testpool/fs@snap2
total estimated size is 0B
root@linux:~# zfs send -v -n -i $POOLNAME/fs#bm1 $POOLNAME/fs@snap2
Error: Unsupported flag with filesystem or bookmark.
root@linux:~# zfs send -v -n -i $POOLNAME/fs@snap1 $POOLNAME/fs#bm2
Error: Unsupported flag with filesystem or bookmark.
root@linux:~# zfs send -v -n -i $POOLNAME/fs#bm1 $POOLNAME/fs#bm2
Error: Unsupported flag with filesystem or bookmark.
```

So i'm adding an explicit error here.

Reply:
I see.  What's the motivation from removing this functionality on the kernel side?  We could move user space over to ZFS_IOC_SEND_SPACE and leave this legacy functionality untouched for compatibility purposes.

Reply:
Just code cleanliness: API consumers should be using libzfs_core anyway, which already leverage ZFS_IOC_SEND_SPACE. `zfs_ioc_send` is lagging behind in terms of functionality (no send-estimates for bookmarks) and will probably never be updated, new code will most likely go to `zfs_ioc_send_new`/`zfs_ioc_send_space`.

I agree it doesn't hurt to leave this here for compatibility though, we can revisit this when we stabilize the API: should i update the PR (both code and commit message) to re-introduce this functionality in `zfs_ioc_send`?

Reply:
Yes, please do.  We want to avoid breaking these legacy interfaces unless we have too, at least for now.  I doubt anyone is using it except for the `zfs` utilities but you never know.  Plus this way the older binaries should continue to work.

I agree we should be transitioning everything to the new interfaces and it would be nice to eventually retire these legacy interfaces.   In fact, let's add a comment to `zfs_ioc_send()` clearly stating this is no longer the preferred interface and `zfs_ioc_recv_new()` should be used.